test-tools/wamr-ide/VSCode-Extension/src/extension.ts

```diff
@@ -461,48 +505,70 @@ export async function activate(context: vscode.ExtensionContext) {
-                detail: 'Can not find WASM binary, please build WASM firstly.',
-            });
-            return;
-        }
-        vscode.commands
-            .executeCommand(
-                'workbench.action.tasks.runTask',
-                'Destroy: Wasm-Container-Before-Run'
-            )
-            .then(() => {
-                const disposableAft = vscode.tasks.onDidEndTaskProcess(e => {
-                    if (e.execution.task.name === 'Wasm-Container-Before-Run') {
-                        /* make sure that run wasm task will be executed after destroy task finish */
-                        vscode.commands
-                            .executeCommand(
-                                'workbench.action.tasks.runTask',
-                                'Run: Wasm'
-                            )
-                            .then(() => {
-                                if (e.exitCode !== 0) {
-                                    disposableAft.dispose();
-                                    return;
-                                }
-                            });
-                        disposableAft.dispose();
-                    }
+            try {
+                /* check if docker images are set up before building */
+                if (
+                    (await checkIfDockerStarted()) &&
+                    !(await checkIfDockerImagesExist(context))
+                ) {
+                    await promptSetupDockerImages(context);
+                }
+            } catch (e) {
+                vscode.window.showWarningMessage((e as Error).message);
+                return;
+            }
+
+            /* refuse to debug if build process failed */
+            if (!checkIfBuildSuccess()) {
+                vscode.window.showErrorMessage('Debug failed', {
+                    modal: true,
+                    detail: 'Can not find WASM binary, please build WASM firstly.',
                 });
-            });
-    });
+                return;
+            }
+
+            vscode.commands
+                .executeCommand(
+                    'workbench.action.tasks.runTask',
+                    'Destroy: Wasm-Container-Before-Run'
+                )
+                .then(() => {
+                    const disposableAft = vscode.tasks.onDidEndTaskProcess(
+                        e => {
+                            if (
+                                e.execution.task.name ===
+                                'Wasm-Container-Before-Run'
+                            ) {
+                                /* make sure that run wasm task will be executed when destroy task finish */
+                                vscode.commands
+                                    .executeCommand(
+                                        'workbench.action.tasks.runTask',
+                                        'Run: Wasm'
+                                    )
+                                    .then(() => {
```

Comment:
Would it be possible to convert this code into `async`? I think this nested code is quite difficult to read and maintain.

Reply:
Thanks so much for taking time to review this PR : )

And yes, the code in this part is not easily to read and may cost unnecessary effort to maintain, I will take it seriously to think about refactoring with `async` and introduce in another patch.